File: app/src/main/java/de/metzgore/beansplan/weeklyschedule/WeeklyScheduleViewModel.java
package de.metzgore.beansplan.weeklyschedule;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;

import de.metzgore.beansplan.data.Resource;
import de.metzgore.beansplan.data.Status;
import de.metzgore.beansplan.data.room.WeeklyScheduleWithDailySchedules;

public class WeeklyScheduleViewModel extends ViewModel {

    private final MutableLiveData<Boolean> refresh = new MutableLiveData<>();
    private final LiveData<Resource<WeeklyScheduleWithDailySchedules>> schedule;
    public final LiveData<Boolean> isEmpty;
    public final LiveData<Boolean> isLoading;

    public WeeklyScheduleViewModel(WeeklyScheduleRepository scheduleRepo) {
        schedule = Transformations.switchMap(refresh, scheduleRepo::loadSchedule);
        isEmpty = Transformations.map(schedule, schedule -> schedule == null || schedule.getData
                () == null || schedule.getData().getDailySchedulesWithShows().isEmpty());
        isLoading = Transformations.map(schedule, schedule -> schedule.getStatus() == Status
                .LOADING);
    }

    public LiveData<Resource<WeeklyScheduleWithDailySchedules>> getSchedule() {
        return schedule;
    }

    public void loadScheduleFromNetwork() {
        refresh.setValue(true);
    }

    public void loadSchedule() {
        refresh.setValue(isEmpty.getValue() == null || isEmpty.getValue());
    }
}

File: app/src/main/java/de/metzgore/beansplan/weeklyschedule/WeeklyScheduleViewModel.kt
package de.metzgore.beansplan.weeklyschedule

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import de.metzgore.beansplan.data.Resource
import de.metzgore.beansplan.data.Status
import de.metzgore.beansplan.data.room.relations.WeeklyScheduleWithDailySchedules

class WeeklyScheduleViewModel(private val scheduleRepo: WeeklyScheduleRepository) : ViewModel() {

    private val refresh = MutableLiveData<Boolean>()
    val schedule: LiveData<Resource<WeeklyScheduleWithDailySchedules>> = Transformations.switchMap(refresh) {
        scheduleRepo.loadSchedule(it)
    }
    val isEmpty: LiveData<Boolean> = Transformations.map(schedule) { schedule ->
        schedule?.data == null || schedule.data.dailySchedulesWithShows.isEmpty()
    }
    val isLoading: LiveData<Boolean> = Transformations.map(schedule) { (status) ->
        status == Status.LOADING
    }

    fun loadScheduleFromNetwork() {
        refresh.value = true
    }

    fun loadSchedule() {
        refresh.value = isEmpty.value == null || isEmpty.value!!
    }
}
